### ISSUE TITLE
✨ Add GRD-Net with optional ROI-aware training and inference

### ISSUE THREAD

**Issue by NickF93** [open-edge-platform/anomalib#3709]
### Problem Description

I would like to contribute **GRD-Net**, a published generative-reconstructive-discriminative anomaly detection model with region-of-interest attention.

GRD-Net combines:

1. a residual encoder-decoder-encoder generator;
2. an adversarial discriminator for feature matching and training stability;
3. a discriminative U-Net that predicts anomaly segmentation maps; and
4. an ROI-aware loss that focuses anomaly supervision on application-relevant regions.

The model is trained on normal images with generated synthetic perturbations. It produces both image-level anomaly scores and pixel-level anomaly maps.

The ROI mechanism addresses an industrial inspection requirement that is not covered by existing anomalib models: defects may only be relevant inside a defined product region, while visually variable areas outside that region should not cause rejection.

Anomalib currently has `gt_mask` for anomaly ground truth, but no general-purpose typed `roi_mask` field. `gt_mask` must not be overloaded because an ROI describes where inspection is relevant, not where an anomaly is present.

### References and ownership

- **Published paper:** [GRD-Net: Generative-Reconstructive-Discriminative Anomaly Detection with Region of Interest Attention Module](https://doi.org/10.1155/2023/7773481)
- **Venue:** *International Journal of Intelligent Systems*, Volume 2023, Article ID 7773481
- **Published:** September 2, 2023
- **Authors:** Niccolò Ferrari, Michele Fraccaroli, and Evelina Lamma
- **Official implementation:** [NickF93/GRD-Net](https://github.com/NickF93/GRD-Net)
- I am Niccolò Ferrari, the first and corresponding author. I wrote the entire GRD-Net codebase and am its sole code owner.
- **Upstream code license:** MIT, copyright © 2024–2026 Niccolò Ferrari
- **MIT relicensing commit:** [`b126fd2`](https://github.com/NickF93/GRD-Net/commit/b126fd2)

### Proposed Solution

I propose adding a public model tentatively named `GRDNet`, implemented as an `AnomalibModule`.

The initial contribution would implement only the **published GRD-Net 2023 PyTorch model**. It would use anomalib's Engine, typed batches, preprocessing, metrics, visualization, configuration, and documentation conventions.

### Proposed initial API

The exact public surface is open to maintainer feedback. An illustrative constructor is:

```python
from anomalib.models import GRDNet

model = GRDNet(
    base_features=64,
    latent_dim=256,
    stages=(2, 2, 2, 2),
    segmentator_base_features=64,
    perlin_probability=0.75,
    adversarial_weight=1.0,
    contextual_weight=50.0,
    encoder_weight=1.0,
    focal_alpha=0.25,
    focal_gamma=2.0,
    generator_lr=1e-4,
    discriminator_lr=1e-4,
    segmentator_lr=1e-4,
)
```

Standard usage without custom ROI masks would remain possible:

```bash
anomalib train \
    --model GRDNet \
    --data anomalib.data.MVTecAD \
    --data.category hazelnut
```

Proposed model behavior:

- **Learning type:** `LearningType.ONE_CLASS`
- **Training:** normal images with synthetic Perlin-based perturbations
- **Model branches:** residual generator, adversarial discriminator, and discriminative U-Net
- **Input:** standard image plus optional ROI mask
- **Output:** standard pixel-level `anomaly_map` and image-level `pred_score`
- **No ROI supplied:** use an all-ones/full-image ROI
- **Pre/post-processing:** anomalib `PreProcessor` and `PostProcessor`
- **Evaluation/visualization:** anomalib `Evaluator` and `Visualizer`
- **Configuration:** explicit jsonargparse-compatible constructor parameters

### ROI data-model decision requested

I would like maintainers to approve how optional ROI information should enter anomalib's typed data flow.

My preferred design is to add optional fields with appropriate validation, transforms, and batching:

```python
roi_mask: Mask | None
roi_mask_path: str | None
```

These would be distinct from:

```python
gt_mask
mask_path
```

Required behavior would include:

- applying the same geometric transforms to the image and ROI;
- nearest-neighbor interpolation for ROI masks;
- preserving ROI masks through item-to-batch collation;
- generating an all-ones ROI when none is supplied;
- documenting that ROI and anomaly ground truth have different semantics.

If maintainers do not want general ROI fields in shared image dataclasses, I would welcome guidance on an alternative typed, model-specific path. I would not use an ad hoc dictionary key or overload `gt_mask`.

### Use ROI only as inference post-processing

The predicted anomaly map could simply be multiplied by an ROI after inference. This would not reproduce the published contribution, because GRD-Net uses ROI information in the discriminative training objective.

### Reuse `gt_mask`

This is not semantically correct. `gt_mask` marks anomalous pixels used for evaluation, while `roi_mask` marks the region where anomalies are relevant. A normal training image can have a valid ROI while having no anomaly ground-truth mask.

### Use an untyped model-specific dictionary key

This would bypass anomalib's dataclass validation, batching, transforms, and public data API. A typed representation is preferable.

### Omit ROI support from the first contribution

The model could be integrated as a generative-reconstructive-discriminative network without ROI support, but that would omit a central published contribution. The all-ones fallback already permits ordinary datasets without requiring ROI files.

### Extend DRÆM

GRD-Net is conceptually related to DRÆM but introduces a residual encoder-decoder-encoder GAN, an adversarial discriminator, different loss orchestration, and ROI-aware discrimination. A separate model provides a clearer public API and attribution boundary.

### Published benchmark results

The following are results reported in the published paper, **not yet an anomalib-port reproduction**.

The main MVTec experiments used 200 training epochs and reported image-level and pixel-level AUROC:

| Category | Image AUROC | Pixel AUROC |
|---|---:|---:|
| Hazelnut | 1.000 | 0.974 |
| Metal nut | 1.000 | 0.962 |
| Pill | 0.985 | 0.958 |
| Cable | 0.995 | 0.981 |

The published industrial BFS-vial meniscus experiment reported, after 30 epochs:

- image AUROC: `0.981`
- pixel AUROC: `0.996`
- accuracy: `0.932`

The proprietary industrial dataset cannot be contributed due to NDA restrictions. Public integration benchmarks will therefore use MVTec AD.

The paper reports:

- learning rate: `1e-4`
- 200 epochs for the main public comparisons
- rotation augmentation in `[-π/2, +π/2]`
- image-level and pixel-level AUROC

The paper does not provide exact GPU/CPU model, wall-clock training time, or VRAM consumption. I do not want to invent those values.

The current upstream PyTorch profile provides these implementation defaults, which are separate from published resource measurements:

- image size: `256×256`
- patch size: `128×128`
- patch stride: `64×64`
- batch size: `8`
- epochs: `200`
- seed: `1337`
- full precision by default
- Python `>=3.11`
- PyTorch `>=2.7`

After architectural approval, I will run the anomalib implementation and report:

- the exact hardware and software environment;
- peak RAM and VRAM;
- per-category training time;
- per-image inference time;
- paper-reported versus anomalib-reproduced results in separate columns;
- seed, preprocessing, patch settings, and evaluation protocol.

### Initial contribution scope

Included:

- GRD-Net 2023 PyTorch architecture
- residual encoder-decoder-encoder generator
- adversarial discriminator
- discriminative U-Net
- synthetic Perlin perturbation
- published generator, discriminator, and focal-loss structure
- optional ROI-aware training and inference
- all-ones ROI fallback
- standard anomalib prediction outputs
- CLI/config integration

Not included:

- DeepIndustrial-SN
- TensorFlow/Keras scaffold
- upstream backend abstraction
- upstream standalone CLI
- custom benchmark runner and CSV reporting system
- proprietary industrial datasets
- unrelated custom dataset adapters
- all paper ablations

### Licensing and provenance

I am the sole owner and author of the upstream GRD-Net codebase. It is now MIT licensed through commit [`b126fd2`](https://github.com/NickF93/GRD-Net/commit/b126fd2).

The paper cites DRÆM, GANomaly, and other methods as conceptual foundations. I authored the upstream implementation and will complete a file-level provenance audit before submitting the PR. Any copied or adapted third-party code identified by that audit will retain its required attribution and notices.

The anomalib contribution would be submitted under anomalib's Apache-2.0 contribution terms. If code is adapted from the upstream MIT implementation, I will preserve:

- the upstream MIT notice and my authorship;
- a colocated `LICENSE`;
- an entry in `third-party-programs.txt`;
- source and publication attribution.

### Planned PR deliverables

After architectural approval, the GRD-Net PR would include:

- `AnomalibModule` integration
- typed ROI data handling approved by maintainers
- public exports and model discovery
- CLI/config example
- unit tests for every model branch and loss
- ROI transformation, batching, and fallback tests
- training, inference, checkpoint, and output-shape tests
- CPU-safe tests without external downloads
- model README and reference documentation
- measured anomalib benchmarks
- sample anomaly maps
- licensing and third-party records
- changelog entry

I will wait for explicit maintainer approval of the ROI data-model design and initial scope before starting the full port.

**Comment by rajeshgangireddy** [comment]
Hi @NickF93 
Thanks for the proposal to adding this second model as well. 
I think from model and license side, it looks good.
My only issue is with adding roi_mask and roi_mask_path to the image dataclass. It would be great to avoid growing the shared schema for a single model.  
So, are there any other ways to introduce these images to the model directly? 

Perhaps you can take a look at other models that use external images (for example EfficientAD uses takes additional images path `imagenet_dir`) . I think something like `roi_dir ` to the model along with the implementation of some helper methods (`prepare_roi_masks()` ?) should be good? 


Another option  would  be to follow the same pattern anomalib already uses for depth data model: a separate _ROIInputFields mixin composed into model-scoped GRDNetItem/GRDNetBatch classes (see DepthItem/_DepthInputFields in anomalib), plus your own dataset variant (mirroring Folder's mask_dir mechanism) and a configure_pre_processor override to carry the ROI mask through the transform pipeline. But this might be a bit too much for a single model ?

**Comment by NickF93** [comment]
Thank you for the suggestions. I agree that introducing model-specific data classes and a dataset variant may be a relatively large addition for a single model.

My preference would nevertheless be the model-scoped typed approach, mainly because ROI masks are spatially paired with individual images. Unlike EfficientAD’s auxiliary ImageNet images, each ROI must remain exactly aligned with its corresponding image through resizing, cropping, flipping, and rotation.

With the simpler `roi_dir` model argument, the image may already have passed through dataset or preprocessor transformations before the model loads its ROI. This should work with deterministic resizing and GRD-Net’s internal augmentations, but arbitrary spatial augmentations in the data pipeline could cause the image and ROI to become misaligned. It would also move per-sample path resolution and file loading into the model.

If we use the typed approach, I would keep its scope narrow:

- model-scoped `_ROIInputFields`, `GRDNetItem`, and `GRDNetBatch`;
- a GRD-Net-specific dataset variant accepting `roi_dir`;
- synchronized transformations for the image, ROI, and anomaly ground-truth mask;
- no changes to the existing shared `ImageItem` or `ImageBatch`;
- an all-ones ROI fallback so ordinary anomalib datasets remain usable.

That said, I understand the maintenance-cost concern and I am happy to follow the approach you prefer. If you would rather keep the first contribution smaller, I can implement the model-level `roi_dir` and `prepare_roi_masks()` approach, document its transformation constraints, and keep all GRD-Net geometric augmentations synchronized inside the model.

Could you please confirm which of these two approaches you would prefer for the initial PR?

**Comment by rajeshgangireddy** [comment]
Your idea of a model-scoped typed approach sounds good but i am not sure how "a GRD-Net-specific dataset variant accepting roi\_dir" would look like in terms of architecture. I suggest that you please go  ahead and create a PR  and we can  provide feedback on the PR itself.

**Comment by andersendsa** [comment]
Hi @NickF93, @rajeshgangireddy  is it okay if i work on the implementation of this issue ?

**Comment by NickF93** [comment]
Hi @andersendsa, thanks for your interest! I’m planning to work on this contribution myself, since GRD-Net is my work and I also maintain the original implementation. I’ve already discussed the main integration choices with @rajeshgangireddy, and the next step is to put together a first PR so we can refine the architecture there.

To avoid working on two separate implementations, perhaps we can coordinate once that PR is open. I’d be happy to have your help with testing, review, or other parts of the integration. Does that work for you?
